### PR TITLE
DDF-1958: Develop reusable Geo Libraries

### PR DESCRIPTION
**Pull request by troymohl**
#### What does this PR do?
Adds generic Geospatial libraries to DDF
#### Who is reviewing it?
@michaelmenousek 
@pklinef 
@shaundmorris 
#### How should this be tested?
Run unit tests
#### Any background context you want to provide?
Start of a place to store generic geospatial libraries.
#### What are the relevant tickets?
DDF-1958
#### Checklist:
- [ ] Documentation Updated
- [X] Update / Add Unit Tests
  Unit tests have been included in test directory.
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: libs/geospatial/src/main/java/org/codice/ddf/libs/geo/conversion/DecimalDegreesConverter.java [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+            setVariables(latitude, longitude);
+
+            int longZone = getLongZone(longitude);
+            LatZones latZones = new LatZones();
+            char latZone = latZones.getLatZone(latitude);
+
+            double easting = getEasting();
+            double northing = getNorthing(latitude);
+
+            // UTM = longZone + " " + latZone + " " + ((int) _easting) + " "
```
Comment (jlcsmith) [inline libs/geospatial/src/main/java/org/codice/ddf/libs/geo/conversion/DecimalDegreesConverter.java]: Can this be removed?

**Comment by bradh** [comment]
This does not implement MGRS correctly. Please check the NGA docs on how polar areas are represented.

File: libs/geospatial/src/main/java/org/codice/ddf/libs/geo/conversion/DecimalDegreesConverter.java [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+        }
+
+        /**
+         * Computes the Longitude zone (UTM Zone)
+         *
+         * @param longitude the logitude value (in decimal degree format) to convert
+         *                  the UTM Zone for.
+         * @return
+         */
+        private int getLongZone(double longitude) {
```
Comment (bradh) [inline libs/geospatial/src/main/java/org/codice/ddf/libs/geo/conversion/DecimalDegreesConverter.java]: This logic does not hold for all zones. There are extended zones in certain latitudes around Norway and Svalbard. See http://earth-info.nga.mil/GandG/publications/tm8358.1/tr83581b.html#ZZ28

File: libs/geospatial/src/main/java/org/codice/ddf/libs/geo/conversion/DecimalDegreesConverter.java [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+         * @return the double value for the UTM easting
+         */
+        private double getEasting() {
+            return 500000 + (k4 * p + k5 * Math.pow(p, 3));
+        }
+
+    }
+
+    /**
+     * Class that represents the UTM LatitudeZones also know as the UTM band
```
Comment (bradh) [inline libs/geospatial/src/main/java/org/codice/ddf/libs/geo/conversion/DecimalDegreesConverter.java]: know -> known.

File: libs/geospatial/src/main/java/org/codice/ddf/libs/geo/conversion/MgrsConverter.java [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+ */
+package org.codice.ddf.libs.geo.conversion;
+
+import org.codice.ddf.libs.geo.coordinate.DecimalDegreesCoordinate;
+import org.codice.ddf.libs.geo.coordinate.DmsCoordinate;
+import org.codice.ddf.libs.geo.coordinate.MgrsCoordinate;
+import org.codice.ddf.libs.geo.coordinate.UtmCoordinate;
+
+/**
+ * Utility class that converts MgrsCoordinate objects to other coordinate
```
Comment (bradh) [inline libs/geospatial/src/main/java/org/codice/ddf/libs/geo/conversion/MgrsConverter.java]: Does not handle UTM limits correctly, or the transform to UPS.

File: libs/geospatial/src/main/java/org/codice/ddf/libs/geo/coordinate/Angle.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+    /**
+     * Calculates the Decimal Degree value for the DMS values
+     * <p>
+     * Note: an IllegalArgumentException will be thrown if any of the values are not in the valid ranges
+     *
+     * @param degrees      the DMS Degree value
+     * @param minutes      the DMS Minute value
+     * @param seconds      the DMS Second value
+     * @param centiseconds the DMS Centisecond value
```
Comment (bradh) [inline libs/geospatial/src/main/java/org/codice/ddf/libs/geo/coordinate/Angle.java]: This probably needs more explanation. Passing a fractional value as an integer as the fraction of a second is really weird API. That is, the legal values are 0 to 999, and its centiseconds, so if I pass 999, its really 9.99 seconds. Can I suggest milliseconds or just allow the seconds to be a floating point value?

File: libs/geospatial/src/main/java/org/codice/ddf/libs/geo/coordinate/Angle.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+     * @param isPositive   boolean value stating whether the coordinate value is positive or negative
```
Comment (bradh) [inline libs/geospatial/src/main/java/org/codice/ddf/libs/geo/coordinate/Angle.java]: This is actually east or west, which would be clearer API.

File: libs/geospatial/src/main/java/org/codice/ddf/libs/geo/coordinate/DmsCoordinate.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    public DmsCoordinate(Latitude lat, Longitude lon) {
+        if (lat == null || lon == null) {
+            throw new IllegalArgumentException("Latitude and Longitude must be non null");
+        }
+        this.latitude = lat;
+        this.longitude = lon;
+    }
+
+    /**
+     * Constructor which converts a String in DMS format to a Latitude/Longitude paise.  The String format is
```
Comment (bradh) [inline libs/geospatial/src/main/java/org/codice/ddf/libs/geo/coordinate/DmsCoordinate.java]: paise -> pair?

File: libs/geospatial/src/main/java/org/codice/ddf/libs/geo/coordinate/DmsCoordinate.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+     * DDMMSS[sss]HDDMMSS[sss]H where (latitude is first):<br/>
+     * <p>
+     * DD is Degrees from 00 to 90 (must always be 2 characters long) representing the latitude
+     * DDD is Degrees from 000 to 180 (must always be 3 characters long) representing longitude<br/>
+     * MM is Minutes from 00 to 59 (must always be 2 characters long)<br/>
+     * SS is Seconds from 00 to 59 (must always be 2 characters long)<br/>
+     * sss is optional centiseconds from 000 to 999 (must either not appear or be 3 characters long)<br/>
+     * H the North, South, East or West Hemisphere representing by the 'N', 'S;, 'E' or 'W' characters respectively
```
Comment (bradh) [inline libs/geospatial/src/main/java/org/codice/ddf/libs/geo/coordinate/DmsCoordinate.java]: quote mark typo.

File: libs/geospatial/src/main/java/org/codice/ddf/libs/geo/coordinate/DmsCoordinate.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+        }
+        this.latitude = lat;
+        this.longitude = lon;
+    }
+
+    /**
+     * Constructor which converts a String in DMS format to a Latitude/Longitude paise.  The String format is
+     * DDMMSS[sss]HDDMMSS[sss]H where (latitude is first):<br/>
+     * <p>
+     * DD is Degrees from 00 to 90 (must always be 2 characters long) representing the latitude
```
Comment (bradh) [inline libs/geospatial/src/main/java/org/codice/ddf/libs/geo/coordinate/DmsCoordinate.java]: Did you want a line break here?

File: libs/geospatial/src/main/java/org/codice/ddf/libs/geo/coordinate/DmsCoordinate.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+     * DDD is Degrees from 000 to 180 (must always be 3 characters long) representing longitude<br/>
+     * MM is Minutes from 00 to 59 (must always be 2 characters long)<br/>
+     * SS is Seconds from 00 to 59 (must always be 2 characters long)<br/>
+     * sss is optional centiseconds from 000 to 999 (must either not appear or be 3 characters long)<br/>
+     * H the North, South, East or West Hemisphere representing by the 'N', 'S;, 'E' or 'W' characters respectively
+     * <p>
+     * <p>
+     * Example of String without centiseconds: 224530N0550230E (approx 22.75 degrees north and 55.40 degrees east)<br/>
```
Comment (bradh) [inline libs/geospatial/src/main/java/org/codice/ddf/libs/geo/coordinate/DmsCoordinate.java]: Longitude conversion looks wrong. You are saying 55 02 30 E is 55.40, but 55.04 might be closer.

File: libs/geospatial/src/main/java/org/codice/ddf/libs/geo/coordinate/DmsCoordinate.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+     *
+     * @param lon the DMS String representation of the Latitude Longitude pair
```
Comment (bradh) [inline libs/geospatial/src/main/java/org/codice/ddf/libs/geo/coordinate/DmsCoordinate.java]: param name is wrong.

File: libs/geospatial/src/main/java/org/codice/ddf/libs/geo/coordinate/Latitude.java [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+ */
+public class Latitude extends Angle {
+
+    private Hemisphere ns;
+
+    /**
+     * Constructor which converts a String in DMS format to a Latitude value.  The String format is
+     * DDMMSS[sss]H where:<br/>
+     * <p>
+     * DD is Degrees from 000 to 90 (must always be 2 characters long)<br/>
```
Comment (bradh) [inline libs/geospatial/src/main/java/org/codice/ddf/libs/geo/coordinate/Latitude.java]: 000 is not two characters.

File: libs/geospatial/src/main/java/org/codice/ddf/libs/geo/coordinate/Latitude.java [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+     * SS is Seconds from 00 to 59 (must always be 2 characters long)<br/>
+     * sss is optional centiseconds from 000 to 999 (must either not appear or be 3 characters long)<br/>
+     * H the North or South Hemisphere representing by the 'N' or 'S' characters respectively
+     * <p>
+     * <p>
+     * Example of String without centiseconds: 0550230E (55 Degrees, 2 minutes and 30 seconds North)<br/>
+     * Example of String with centiseconds: 0550230789E (55 Degrees, 2 minutes and 30 seconds and 789 centiseconds North)
+     *
+     * @param lat the DMS String representation of the Longitude
+     * @throws GeoFormatException if the longitude String is not in the valid format and cannot be parsed
```
Comment (bradh) [inline libs/geospatial/src/main/java/org/codice/ddf/libs/geo/coordinate/Latitude.java]: latitude String

**Comment by bradh** [comment]
There should be edge-condition tests. For example, try DMS outside the expected valid range, or polar regions UTM/MGRS.

**Comment by bradh** [comment]
There are more variations than just decimal degrees and DMS. One common example is DDMM.mmm (e.g. ACTDF does this). I can accept that DD and DMS works for you now, but there are API choices that don't give you much extension room. It might be worth having some static factory methods in place of constructors.

**Comment by bradh** [comment]
This code has excessive functionality duplication with the org.geotools.measure code. Perhaps you could have extended / wrapped that code instead?

**Comment by troymohl** [comment]
Going to refactor this to wrap the Geotools APIs. Also will remove MGRS and UTM until they are needed.

**Pull request closed without merging**